scripts/esbuild/cli.ts

```diff
@@ -0,0 +1,75 @@
+import type { BuildOptions as ESBuildOptions } from 'esbuild';
+import * as esbuild from 'esbuild';
+import fs from 'fs-extra';
+import { join } from 'path';
+
+import { BuildOptions } from '../utils/options';
+import { writePkgJson } from '../utils/write-pkg-json';
+import { getEsbuildAliases, getEsbuildExternalModules } from './util';
+
+/**
+ * Runs esbuild to bundle the `cli` submodule
+ *
+ * @param opts build options needed to generate the rollup configuration
+ * @returns an array containing the generated rollup options
+ */
+export async function buildCli(opts: BuildOptions) {
+  // clear out rollup stuff
+  fs.emptyDir(opts.output.cliDir);
+
+  const inputDir = join(opts.srcDir, 'cli');
+  const buildDir = join(opts.buildDir, 'cli');
+
+  const outputDir = opts.output.cliDir;
+  const esmFilename = 'index.js';
+  const cjsFilename = 'index.cjs';
+
+  const dtsFilename = 'index.d.ts';
+
+  const cliAliases = getEsbuildAliases(opts);
+
+  const external = getEsbuildExternalModules(opts, opts.output.cliDir);
+
+  const cliEsbuildOptions: ESBuildOptions = {
+    entryPoints: [join(inputDir, 'index.ts')],
+    bundle: true,
+    platform: 'node',
+    logLevel: 'info',
+    sourcemap: 'linked',
+    external,
+    alias: cliAliases,
+  };
+
+  // ESM build
+  await esbuild.build({
+    ...cliEsbuildOptions,
+    outfile: join(outputDir, esmFilename),
+    format: 'esm',
+  });
+
+  // CommonJS build
+  await esbuild.build({
+    ...cliEsbuildOptions,
+    outfile: join(outputDir, cjsFilename),
+    format: 'cjs',
+  });
```

Comment:
Do these need to run sequentially, or can they run in parallel? i.e. using `await Promise.all(...)`

Reply:
oh yeah they could be in parallel I think, although I don't know what the performance implications of that would be. Esbuild is implemented in Go so the call to `esbuild.build` is actually going to call out to a native node extension which will do all the hard work (there's basically a small JS shim around the Go code afaik). The extension being written in Go means that it has its own internal threads and whatnot - I think it will basically saturate your CPU provided the job takes long enough. It could even be that if you try to fire off two esbuild processes at the same time they are slightly slower because you get some resource contention between them, but in this case the execution times are so brief that who knows.

If we want to we could do some empirical testing to get an idea of whether we see a significant difference from using `Promise.all` here